File: js/euler58.js
function euler58(){
    // 3, 5, 7, 13, 17, 31, 37, 43
    // 8 primes out of 13 numbers on diagonals of 7x7 square spiral
    var side = 7;
    var primes = 8;
    // continue while primes make up at least 10% of numbers on diagonals
    while ((2 * side - 1) < 10 * primes){
        side += 2;
        // new numbers on corners of square spiral of side length n
        // n^2, n^2 - n + 1, n^2 - 2n + 2, n^2 - 3n + 3
        for (let i = 0; i <= 3; i++){
            if (isPrime(side * side - i * (side - 1))) primes++;
        }
    }
    return side;
}

function isPrime(n){
    for (let i = 2; i * i <= n; i++){
        if (n % i == 0) return false;
    }
    return true;
}

console.log(euler58());

File: go/euler58.go
package main
import "fmt"

func isPrime(n int) bool {
    if n < 2 {
        return false
    }
    for i := 2; i * i <= n; i++ {
        if n % i == 0 {
            return false
        }
    }
    return true
}

func main() {
    res := 0
    // square spiral with side length 7
    // has 13 numbers along both diagonals, 8 are prime
    primes, count:= 8, 13
    // adding another layer increases side length by 2
    // and adds 4 numbers to diagonals, corners of square
    for n := 9; res == 0; n += 2 {
        // check primality of corner numbers
        // n^2, n^2 - n + 1, n^2 - 2n + 2, n^2 - 3n + 3
        for k := 0; k < 4; k++ {
            corner := n * n - k * (n - 1)
            if isPrime(corner) {
                primes++
            }
            count++
        }
        // return 
        if primes * 10 < count {
            res = n
        }
    }
    fmt.Println(res)
}
